### ISSUE TITLE
Version 1.4.0 - Error

### ISSUE THREAD

**Issue by neuroserve** [arekinath/triton-dehydrated#52]
dehydrated 1.4.0 errors out. I guess, there's a typo somewhere, but I didn't find it:

```
[root@myinstance /opt/letsencrypt]# /opt/letsencrypt/dehydrated --register --accept-terms
# INFO: Using main config file /opt/letsencrypt/config
egrep: illegal option -- a
usage: egrep [-bchHilLnoqrRsvx] [-A num] [-B num] [-C num|-num]
             [--label=name] [-e pattern_list]... [-f pattern_file]...
             [pattern_list] [file]...
EXPECTED value GOT EOF
```

**Comment by danmcd** [comment]
Oh wow, another grep option we need to pull in.  From MacOS X:

```
     -a, --text
             Treat all files as ASCII text.  Normally grep will simply print
             ``Binary file ... matches'' if files contain binary characters.
             Use of this option forces grep to output lines matching the spec-
             ified pattern.
```

illumos grep and egrep do not have -a at the moment.  See https://www.illumos.org/issues/4114 for why it might be okay for use to remove the "-a" from dehydrated on illumos until such time as we implement illumos#4114.

**Comment by bahamat** [comment]
The grep error can be ignored. The actual error you're getting is `EXPECTED value GOT EOF`. That does sound familiar...I'll do some testing.

**Comment by bahamat** [comment]
@neuroserve Check this issue: dehydrated-io/dehydrated#763

I think maybe you have the v1 api endpoint in your config. Let's Encrypt doesn't provide the v1 API anymore.

**Comment by neuroserve** [comment]
I've just untarred the 1.4.0 triton-dehydrated release tarball and changed "domain.txt". Nothing more. I'm not even sure how I would change the api endpoint. Using version 1.3.2 everything works without a flaw.

The config files from both version tarballs are the same - no difference. The error can even be seen without using a config file, at all.

This takes place in a minimal-64-lts@20.4.0 (image: 800db35c-5408-11eb-9792-872f658e7911), where I have done "pkgin update", "pkgin upgrade" and installed nginx 1.21.0.

**Comment by bahamat** [comment]
OK, I see. Usually we have a `config.overrides` file which sets the API. I guess dehydrated defaults to the v1 endpoint if nothing is set.

To work around this, you can uncommon the `CA` directive in `config` until I get a new version published next week.

**Comment by neuroserve** [comment]
I'm not sure about the v1 default. This is in deps/dehydrated/dehydrated:

`# Preset
CA_ZEROSSL="https://acme.zerossl.com/v2/DV90"
CA_LETSENCRYPT="https://acme-v02.api.letsencrypt.org/directory"
CA_LETSENCRYPT_TEST="https://acme-staging-v02.api.letsencrypt.org/directory"
CA_BUYPASS="https://api.buypass.com/acme/directory"
CA_BUYPASS_TEST="https://api.test4.buypass.no/acme/directory"
`
The only reference to APIv1 is in the changelog for version 0.6.5 - but triton-dehydrated 1.3.2 is already using 0.6.5. No mentioning of APIv1 in the changelog regarding newer versions or even 0.7.1, which is in use in triton-dehydrated 1.4.0.

**Comment by bahamat** [comment]
I'm not able to reproduce what you describe.

```
# mkdir -p /opt/dehydrated
# latest=$(curl -s https://api.github.com/repos/joyent/triton-dehydrated/releases/latest | json assets.0.browser_download_url)
# echo $latest
https://github.com/joyent/triton-dehydrated/releases/download/v1.4.0/dehydrated.tar.gz
# curl -L "$latest" | gtar --no-same-owner -zx -C /opt/dehydrated
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   623  100   623    0     0  77875      0 --:--:-- --:--:-- --:--:-- 77875
100   98k  100   98k    0     0  1261k      0 --:--:-- --:--:-- --:--:-- 1261k
# cd /opt/dehydrated/
# echo test.joyent.com > domains.txt
# ./dehydrated --register --accept-terms
# INFO: Using main config file /opt/dehydrated/config
+ Generating account key...
+ Registering account key with ACME server...
+ Fetching account URL...
+ Done!
```

In order to debug further, you'll need to edit `deps/dehydrated/dehydrated`. Find the `jsonsh()` function. Inside of that, there's a `throw()` function.

Change the `throw()` function to this:

```
throw() {
    set +o xtrace

    (( cnt = ${#FUNCNAME[@]} ))
    (( i = 0 ))
    while (( i < cnt )); do
        printf '  [%3d] %s\n' "${i}" "${FUNCNAME[i]}"
        if (( i > 0 )); then
            line="${BASH_LINENO[$((i - 1))]}"
        else
            line="${LINENO}"
        fi
        printf '        (file "%s" line %d)\n' "${BASH_SOURCE[i]}" "${line}"
        (( i++ ))
    done
    exit 1
}
```

Then try it again and send me the output.

**Comment by neuroserve** [comment]
Just to be sure, I have created a new test instance - without upgrading pkgin, etc.:
```
[root@dehydrate /opt/root/letsencrypt]# latest=$(curl -s https://api.github.com/repos/joyent/triton-dehydrated/releases/latest | json assets.0.browser_download_url)
[root@dehydrate /opt/root/letsencrypt]# echo $latest
https://github.com/joyent/triton-dehydrated/releases/download/v1.4.0/dehydrated.tar.gz
[root@dehydrate /opt/root/letsencrypt]# curl -L "$latest" | gtar --no-same-owner -zx -C /opt/root/letsencrypt
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   623  100   623    0     0   1780      0 --:--:-- --:--:-- --:--:--  2170
100   98k  100   98k    0     0   166k      0 --:--:-- --:--:-- --:--:--  166k
[root@dehydrate /opt/root/letsencrypt]# cd /opt/root/letsencrypt/
[root@dehydrate /opt/root/letsencrypt]# echo test.code667.net > domains.txt
[root@dehydrate /opt/root/letsencrypt]# ./dehydrated --register --accept-terms
# INFO: Using main config file /opt/root/letsencrypt/config
egrep: illegal option -- a
usage: egrep [-bchHilLnoqrRsvx] [-A num] [-B num] [-C num|-num]
             [--label=name] [-e pattern_list]... [-f pattern_file]...
             [pattern_list] [file]...
EXPECTED value GOT EOF

```
Having changed throw() inside of jsonsh() and executing again:

```
[root@dehydrate /opt/root/letsencrypt]# ./dehydrated --register --accept-terms
# INFO: Using main config file /opt/root/letsencrypt/config
egrep: illegal option -- a
usage: egrep [-bchHilLnoqrRsvx] [-A num] [-B num] [-C num|-num]
             [--label=name] [-e pattern_list]... [-f pattern_file]...
             [pattern_list] [file]...
```
BTW.: I've just checked with an instance with only one network interface - same behaviour.

**Comment by bahamat** [comment]
OK, I've finally tracked this down. The fix is in dehydrated-io/dehydrated#840. As soon as it's merged upstream, I'll update triton-dehydrated.

In the mean time, if necessary you can hotpatch your dehydrated dependency.

**Comment by bahamat** [comment]
This is fixed in the 1.4.2 release bundle.
